### PR TITLE
Moving wink light and hub to new model

### PR DESCRIPTION
**Pull request by danielrischar**
Converting wink light to the new resource composition model

### CODE REVIEW

File: opent2t.p.light/opent2t.p.light.raml [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        body:
+          application/json:
+            schema: Dimming
+            example: |
+              {
+                "id": "unique_example_id",
+                "dimmingSetting": 50
+              }
+
+/SubscribeLampResURI:
```
Comment (arjun-msft) [inline opent2t.p.light/opent2t.p.light.raml]: >SubscribeLampResURI [](start = 1, length = 19)

what gets subscribed to? resource or platform?

File: opent2t.p.light/com.wink.lightbulb/js/manifest.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  <voice id="org.opent2t.sample.lightcontrol.cortana.vcd.superpopular" />
+  <!-- associated schemas -->
+  <schemas>
+    <schema id="opent2t.p.light" main="true" type="OCF" />
+    <schema id="oic.core" type="OCF" />
+    <schema id="oic.baseresource" type="OCF" />
+    <schema id="oic.r.switch.binary" type="OCF" />
+    <schema id="oic.r.light.dimming" type="OCF" />
+    <schema id="oic.r.colour.rgb" type="OCF" />
+  </schemas>
```
Comment (begillil) [inline opent2t.p.light/com.wink.lightbulb/js/manifest.xml]: Missing a reference to the new opent2t.platform schema, as well as the oic.oic-link schema, and oic.r.colour.chroma.

File: opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+/**
+ * 
+ */
+function providerSchemaToPlatformSchema(providerSchema, expand) {
+    var stateReader = new StateReader(providerSchema.desired_state, providerSchema.last_reading);
+
+    // Build the oic.r.switch.binary resource
+    var binarySwitch = {
```
Comment (begillil) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: You may want to consider naming these variables after the name of the resource instead of the type of the resource, otherwise the precedent gets hard to follow when we have multiple resources of the same type that we need to deal with.

File: opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        dimming.id = 'dim';
+        dimming.dimmingSetting = scaleDeviceBrightnessToTranslatorBrightness(stateReader.get('brightness'));
+        dimming.range = [0,100];
+    }
+
+    return {
+        opent2t: {
+            schema: 'opent2t.p.light',
+            translator: 'opent2t-translator-com-wink-lightbulb',
+            control_id: providerSchema['object_type'] + '.' + providerSchema['object_id']
```
Comment (begillil) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: I think this should just be the object_id, not the concatenated type.id

File: opent2t.p.light/opent2t.p.light.raml [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+                        "dimmingSetting": 30,
+                        "step": 5,
+                        "range": [0, 100]
+                      }
+                    ]
+                  }
+                ]
+              }
+
+/device/{deviceid]/power:
```
Comment (begillil) [inline opent2t.p.light/opent2t.p.light.raml]: Each of these resources can be exposed relative to a single /devices/{deviceid} so we don't have to duplicate the full path every time, e.g.:

```raml
/devices/{deviceid}:
  /power:
    displayName: Power
    ...

  /dim:
    displayName: Dim
    ...
```

Also note: There's a typo in all of the resource paths. You have a ] instead of }

File: opent2t.p.light/opent2t.p.light.raml [hunk 6]
```diff
@@ -0,0 +1,381 @@
+#%RAML 0.8
+title: org.opent2t.sample.lamp.superpopular
```
Comment (begillil) [inline opent2t.p.light/opent2t.p.light.raml]: Needs to be updated to match the new name

File: opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        dim.id = 'dim';
+        dim.dimmingSetting = scaleDeviceBrightnessToTranslatorBrightness(stateReader.get('brightness'));
+        dim.range = [0,100];
+    }
+
+    return {
+        opent2t: {
+            schema: 'opent2t.p.light',
+            translator: 'opent2t-translator-com-wink-lightbulb',
+            control_id: providerSchema['object_id']
```
Comment (begillil) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: I think we've been using camelCase instead of underscores for other JSON field names, we probably want to stick with that, so controlId instead of control_id.

File: opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        return winkHub.putDeviceDetailsAsync(deviceType, deviceId, putPayload)
+            .then((response) => {
+                var schema = providerSchemaToPlatformSchema(response.data, true);
+
+                return findResource(schema, di, resourceId);
+            });
+    }
+
+    getLDevicesPower(deviceId) {
+        return this.getDeviceResource(deviceId, "oic.r.switch.binary");
```
Comment (begillil) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: For this, and other uses of get/postDeviceResource, you should be looking up resources by their id, not their resource type, so for this call it would be 'power', not 'oic.r.switch.binary'.

For this Light platform, there are no duplicated resource types, so it would be fine in theory to look up resources by their type, but that wouldn't work for other platforms, like a 'thermostat' that has multiple 'oic.r.temperature' resources.

File: opent2t.p.light/opent2t.p.light.raml (hunk 1 above)
Comment (arjun-msft) [inline opent2t.p.light/opent2t.p.light.raml]: lets change this from the old name to just /subscribe

---
In reply to: [87448818](https://github.com/openT2T/translators/pull/117#discussion_r87448818) [](ancestors = 87448818)

File: opent2t.p.light/com.wink.lightbulb/js/package.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        "files": [
+            "tests/*.js"
+        ]
+    },
+    "dependencies": {
+        "opent2t-translator-com-wink-hub": "*"
+    },
+    "devDependencies": {
+        "optimist": "*",
+        "opent2t": "*",
```
Comment (arjun-msft) [inline opent2t.p.light/com.wink.lightbulb/js/package.json]: >opent2t [](start = 9, length = 7)

why is this a devDependency?

File: opent2t.p.light/com.wink.lightbulb/js/package.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        "opent2t-translator-com-wink-hub": "*",
```
Comment (arjun-msft) [inline opent2t.p.light/com.wink.lightbulb/js/package.json]: >opent2t [](start = 9, length = 7)

you don't need this in both dev and non

File: opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    // Find the resource
+    var resource = entity.resources.find((r) => {
+        return r.id === resourceId;
+    });
+
+    return resource;
+}
+
+/***
+ *
```
Comment (arjun-msft) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: add comment

Comment (arjun-msft) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: to every method, not just this one

---
In reply to: [87662992](https://github.com/openT2T/translators/pull/117#discussion_r87662992) [](ancestors = 87662992)

File: opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        },
+        pi: providerSchema['uuid'],
+        mnmn: providerSchema['device_manufacturer'],
+        mnmo: providerSchema['manufacturer_device_model'],
+        n: providerSchema['name'],
+        rt: ['opent2t.p.light'],
+        entities: [
+            {
+                rt: ['opent2t.d.light'],
+                di: 'F8CFB903-58BB-4753-97E0-72BD7DBC7933',
```
Comment (arjun-msft) [inline opent2t.p.light/com.wink.lightbulb/js/thingTranslator.js]: >F8CFB903 [](start = 21, length = 8)

what is this magic guid?

File: opent2t.p.light/com.wink.lightbulb/js/tests/test.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
+const sleep = require('es6-sleep').promise;
+var test = require('ava');
+var OpenT2T = require('opent2t').OpenT2T;
+var config = require('./testConfig');
+
+console.log("Config:");
+console.log(JSON.stringify(config, null, 2));
+
+var translatorPath = require('path').join(__dirname, '..');
+var hubPath = require('path').join(__dirname, '../../../../org.opent2t.sample.hub.superpopular/com.wink.hub/js');
```
Comment (arjun-msft) [inline opent2t.p.light/com.wink.lightbulb/js/tests/test.js]: remember to change this path when you change the schema for hub

**Comment by arjun-msft** [comment]
:clock1:

File: opent2t.p.hub/com.wink.hub/js/package.json [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  "bugs": {
+    "url": "https://github.com/openT2T/translators/issues"
+  },
+  "homepage": "https://github.com/openT2T/translators#readme",
+  "dependencies": {
+    "request": "^2.74.0",
+    "request-promise": "^4.1.1",
+    "opent2t": "^1.0.1",
+    "opent2t-onboarding-org-opent2t-onboarding-winkhub": "^0.0.1",
+    "opent2t-translator-com-wink-lightbulb": "*",
```
Comment (arjun-msft) [inline opent2t.p.hub/com.wink.hub/js/package.json]: add min version numbers, use the current translator version and +.0.0.1 it so the current one doesn't get used by accident

File: opent2t.p.hub/com.wink.hub/js/thingTranslator.js [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+        this._baseUrl = "https://api.wink.com";
+        this._devicesPath = '/users/me/wink_devices';
+
+        this._name = "Wink Hub"; // TODO: Can be pulled from OpenT2T global constants. This information is not available, at least, on wink hub.
+    }
+
+    /**
+     * Get the list of devices discovered through the hub.
+     */
+    getPlatforms(expand) {
```
Comment (arjun-msft) [inline opent2t.p.hub/com.wink.hub/js/thingTranslator.js]: shouldn't there also just be a get(expand)?

File: opent2t.p.hub/com.wink.hub/js/thingTranslator.js [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+     * Given the hub specific device, returns the opent2t schema and translator
+    */
+    _getOpent2tInfo(winkDevice) {
+        // if (winkDevice.thermostat_id) {
+        //     return { 
+        //         "schema": 'opent2t.p.thermostat',
+        //         "translator": "opent2t-translator-com-wink-thermostat"
+        //     };
+        // }
+        // else if (winkDevice.binary_switch_id) {
```
Comment (arjun-msft) [inline opent2t.p.hub/com.wink.hub/js/thingTranslator.js]: uncomment this since it has already been merged.

File: opent2t.p.hub/com.wink.hub/js/thingTranslator.js [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+        return this._makeRequest(requestPath, 'GET').then((response) => {
+            return response.data;
+        });
+    }
+
+    /** 
+     * Given the hub specific device, returns the opent2t schema and translator
+    */
+    _getOpent2tInfo(winkDevice) {
+        // if (winkDevice.thermostat_id) {
```
Comment (arjun-msft) [inline opent2t.p.hub/com.wink.hub/js/thingTranslator.js]: depending on who (you or Ben) gets merged second will uncomment this change

File: opent2t.p.hub/opent2t.p.hub.raml [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+schemas:
+  - Platform: !include opent2t.p.hub.json
+
+traits:
+  - interface-sensor:
+      queryParameters:
+        if:
+          enum: ["oic.if.s", "oic.if.baseline"]
+
+/platforms:
```
Comment (arjun-msft) [inline opent2t.p.hub/opent2t.p.hub.raml]: I *think* this is good, we should also have a default get as well.

Comment (danielrischar) [inline opent2t.p.hub/opent2t.p.hub.raml]: #Resolved Added default get and get(expand)

---
In reply to: [88085084](https://github.com/openT2T/translators/pull/117#discussion_r88085084) [](ancestors = 88085084)
